### PR TITLE
Introduce BoxOneOf for Kotlin

### PR DESCRIPTION
**Pull request by oldergod**
First commit is code generation, second commit is the logic changes + tests.

Address #1448 for Kotlin.

The idea is that really big proto message usually contains `oneofs` which have way too many fields and so we introduced the concept of boxed oneof.

### CODE REVIEW

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 1]
```diff
@@ -391,14 +393,23 @@ class KotlinGenerator private constructor(
...
                 newName(field.name, field)
               }
             }
+            message.boxOneOfs().forEach { oneOf ->
+              newName(oneOf.name, oneOf)
+              newName(oneOf.name + "Keys",oneOf.name + "Keys")
+              newName(oneOf.name.capitalize(), oneOf.name.capitalize())
+               oneOf.fields.forEach { field ->
+                 newName(oneOf.name + field.name.capitalize(), oneOf.name + field.name.capitalize())
+               }
```
Comment (swankjesse) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:411]: I think we want a little indirection model here, so we can preserve declaration order.

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 2]
```diff
@@ -464,10 +475,16 @@ class KotlinGenerator private constructor(
 
     addMessageConstructor(type, classBuilder)
 
-    if (type.oneOfs.isNotEmpty()) {
+    if (type.flatOneOfs().isNotEmpty()) {
       classBuilder.addInitializerBlock(generateInitializerOneOfBlock(type))
```
Comment (swankjesse) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:479]: Rename to `generateInitializerFlatOneOfBlock(type)` ?

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 3]
```diff
@@ -864,6 +957,29 @@ class KotlinGenerator private constructor(
           .build())
     }
 
+    val boxOneOfs = message.boxOneOfs()
+    for (oneOf in boxOneOfs) {
+      val fieldClass = message.oneOfClassFor(oneOf, nameAllocator)
+      // Name allocator
```
Comment (swankjesse) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:963]: What’s this comment mean?

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 4]
```diff
@@ -966,6 +1079,18 @@ class KotlinGenerator private constructor(
             }
           })
         }
+        for (oneOf in boxOneOfs) {
+          val fieldName = localNameAllocator[oneOf]
+          add("if (%N != null) ", fieldName)
+          addStatement("%N += %P", resultName, buildCodeBlock {
+            add(fieldName)
+            // TODO(Benoit) Do we redact if one of the field is redacted?
```
Comment (swankjesse) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:1087]: Do we have a tracking bug for these oneofs? Useful to put this in the checklist on that bug.

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 5]
```diff
@@ -1299,13 +1465,14 @@ class KotlinGenerator private constructor(
...
     val redactedFields = mutableListOf<CodeBlock>()
-    for (field in message.fieldsAndOneOfFields) {
+    for (field in message.fieldsAndFlatOneOfFields()) {
       val fieldName = nameAllocator[field]
       val redactedField = field.redact(fieldName)
       if (redactedField != null) {
         redactedFields += CodeBlock.of("%N = %L", fieldName, redactedField)
       }
     }
+    // TODO(benoit) How do we deal with redacted oneof fields !?
```
Comment (swankjesse) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:1475]: For our checklist!

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 6]
```diff
@@ -1776,6 +1943,145 @@ class KotlinGenerator private constructor(
     }
   }
 
+  private fun oneOfBoxType(boxClassName: ClassName, oneOf: OneOf): TypeSpec {
```
Comment (swankjesse) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:1946]: I really like a pasted sample output

File: wire-library/wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/Form.kt [hunk 7]
```diff
@@ -106,79 +42,60 @@ public class Form(
...
+    choice: OneOf<Choice<*>, *>? = this.choice,
+    decision: OneOf<Decision<*>, *>? = this.decision,
     unknownFields: ByteString = this.unknownFields
-  ): Form = Form(button_element, local_image_element, remote_image_element, money_element,
-      spacer_element, text_element, customized_card_element, address_element, text_input_element,
-      option_picker_element, detail_row_element, currency_conversion_flags_element, unknownFields)
+  ): Form = Form(choice, decision, unknownFields)
+
+  /**
+   *
```
Comment (swankjesse) [inline wire-library/wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/Form.kt:75]: Ick.

File: wire-library/wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/Form.kt [hunk 8]
```diff
@@ -191,107 +108,126 @@ public class Form(
...
-                CurrencyConversionFlagsElement.ADAPTER.decode(reader)
-            else -> reader.readUnknownField(tag)
+            else -> {
+              for (choiceKey in choiceKeys) {
+                if (tag == choiceKey.tag) {
+                  choice = choiceKey.decode(reader)
+                  return@forEachTag Unit
+                }
+              }
+              reader.readUnknownField(tag)
```
Comment (swankjesse) [inline wire-library/wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/Form.kt:134]: bug!

File: wire-library/wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/Form.kt [hunk 9]
```diff
@@ -191,107 +108,126 @@ public class Form(
...
-        option_picker_element =
-            value.option_picker_element?.let(OptionPickerElement.ADAPTER::redact),
-        detail_row_element = value.detail_row_element?.let(DetailRowElement.ADAPTER::redact),
-        currency_conversion_flags_element =
-            value.currency_conversion_flags_element?.let(CurrencyConversionFlagsElement.ADAPTER::redact),
         unknownFields = ByteString.EMPTY
       )
     }
 
+    public val choiceButton_element: Choice<ButtonElement> = Choice<ButtonElement>(tag = 1, adapter
```
Comment (swankjesse) [inline wire-library/wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/Form.kt:157]: Should be CHOICE_BUTTON_ELEMENT

File: wire-library/wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/Form.kt [hunk 10]
```diff
@@ -191,107 +108,126 @@ public class Form(
...
+        "option_picker_element")
+
+    public val choiceDetail_row_element: Choice<DetailRowElement> = Choice<DetailRowElement>(tag =
+        11, adapter = DetailRowElement.ADAPTER, declaredName = "detail_row_element")
+
+    public val choiceCurrency_conversion_flags_element: Choice<CurrencyConversionFlagsElement> =
+        Choice<CurrencyConversionFlagsElement>(tag = 12, adapter =
+        CurrencyConversionFlagsElement.ADAPTER, declaredName = "currency_conversion_flags_element")
+
+    public val choiceKeys: Set<Choice<*>> = setOf(choiceButton_element, choiceLocal_image_element,
```
Comment (swankjesse) [inline wire-library/wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/Form.kt:198]: CHOICE_KEYS

**Review by swankjesse — APPROVED** [review]

**Comment by JakeWharton** [comment]
I don't understand why this landed. The problem in the linked issue used oneof's but is easily reproducible with regular fields as well. There should be no special casing of oneof's needed for the fix to that issue.

**Comment by oldergod** [comment]
That's true.
The approach is of 2 origins:
1. the first assumption is that a message with that many fields is highly likely using oneof to endup in such a state.
2. Instead of fixing every potential problematic methods in the generated, we wanted to do something smart and so we went with this approach of boxed oneofs which can be seen as the Wire's analog version to the Any type.

**Comment by swankjesse** [comment]
(if I have a message with a very large number of fields, will it work with any other programming language?)

**Pull request merged**